crates/circuit/src/dag_circuit.rs

```diff
@@ -2216,6 +2233,182 @@ impl DAGCircuit {
         })
     }
 
+    /// Are these two DAGs structurally equal?
+    ///
+    /// This function returns true iff the graph structures are precisely the same as each other,
+    /// including the valid node indices, edge orders, and so on.  This is a much stricter check
+    /// than graph equivalence, and is mostly useful for testing if two :class:`DAGCircuit`
+    /// instances have been constructed and manipulated in the exact same ways.  For example, this
+    /// method can be used to test whether a sequence of manipulations of a DAG is deterministic.
```

Comment:
I might leave a note here that refers people to `__eq__` if they want to do a more general equality test and that this is very specifically for checking determinism.

Reply:
Done (basically) in e28c8d225a.  I think a larger rework of the documentation should go into it more, which is one of our oldest open issues: #3404.